lighthouse-core/gather/gatherers/seo/font-size.js

```diff
@@ -240,25 +221,34 @@ class FontSize extends Gatherer {
   }
 
   /**
-   * Maps backendNodeId of TextNodes to {fontSize, textLength}.
-   *
+   * Iterates on the TextNodes in a DOM Snapshot.
+   * Every entry is associated with a TextNode in the layout tree (not display: none).
    * @param {LH.Crdp.DOMSnapshot.CaptureSnapshotResponse} snapshot
-   * @return {BackendIdsToFontData}
    */
-  calculateBackendIdsToFontData(snapshot) {
+  * iterateTextNodesInLayoutFromSnapshot(snapshot) {
```

Comment:
> outside of the very unusual asset saver case which is very specific to allowing streamable results

(and really is a product of the Node < 8.10 days when we couldn't have strings over 256MiB (https://github.com/GoogleChrome/lighthouse/issues/1685#issuecomment-279115446) and no one has touched that save method for three years)

Reply:
Happy to change, not a sticking point, but I don't like the idea of avoiding entire language features because they are new or not commonly used :) Would a comment linking to "how generators work" help? Is the main concern just that someone doesn't know what `*` is, or that even if they do, do you think generators are hard to reason about?

I appreciate a generator here as the size of the data created is not bounded. It scales with the number of elements on the page, and with the amount of text within those elements. Since only a single value needs to be processed at a time, and because the creation of said value is rather complex, extracting as a generator function made sense to me.

The alternative seems to be:

1. inline this entire function. affects readability, function decomposition is good, etc.
2. return as an array. this is the current state, and I believe I reduced some of the concern here in a previous PR by returning the `textLength` instead of the `text`. Here I reverted that, as it's a derived property better calculated by the calling code.

Reply:
>  Is the main concern just that someone doesn't know what * is, or that even if they do, do you think generators are hard to reason about?

Mostly this. I wouldn't expect most developers to immediately understand how this function works compared to adding items to an array. For this reason, I would only really use a generator in a shared codebase when the benefits are *significant* (like in the string creation case of asset saver) which is accompanied by that justification in comments.

> I believe I reduced some of the concern here in a previous PR by returning the textLength instead of the text. Here I reverted that, as it's a derived property better calculated by the calling code.

Ah, OK I didn't notice this. If we held on to `text` for any significant part of the filtering step I might agree with you this is sufficient justification, but much like your other simplifications in this PR, we could solve it by limiting the returned information and just change `text` to a `textLength: getTextLength(text)` invocation into the iterator function, right? 

> as it's a derived property better calculated by the calling code.

I'm not sure I follow this one, we'd still be calculating it with the exact same method now, we'd just compute it before returning the node object.

Final thoughts:

I understand your concern about refusing to adopt new language features and ending up with a ES20XX-era codebase forever. I'm not as worried about this being the case as we've extensively adopted several features newer to node than generators, it's primarily that generators are rarer in applicability for Lighthouse (and therefore in my belief less likely to be easy to read for the average developer) than asynchronous code.

While my primary concern was wrong to lean exclusively on "it's different, so no", there is still *some* merit to consistency in patterns. @paulirish mostly converted me here on avoiding `reduce` in pretty much every situation I would've used it previously for easier reading to those less accustomed to `reduce` as a general-purpose `for`-loop, and I believe in other recent reviews of new contributors we both concurred on use of a `for...of` instead of a long `.filter().map().filter().forEach()` chain. Not that there's never a time to use those new patterns, but for a situation in which the dominant pattern already fits well and the new pattern is sufficiently different in complexity of comprehension, I'd prefer the former.